blockstack-vm/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+pub mod types;
+pub mod representations;
+pub mod parser;
+pub mod contexts;
+pub mod errors;
+pub mod database;
+
+mod functions;
+
+use types::{Value, CallableType};
+use representations::SymbolicExpression;
+use contexts::{Context, Environment};
+use functions::define::DefineResult;
+use errors::Error;
+
+type InterpreterResult = Result<Value, Error>;
+
+fn lookup_variable(name: &str, context: &Context, env: &Environment) -> InterpreterResult {
+    // first off, are we talking about a constant?
+    if name.starts_with(char::is_numeric) {
+        match i128::from_str_radix(name, 10) {
+            Ok(parsed) => Ok(Value::Int(parsed)),
+            Err(_e) => Err(Error::Generic("Failed to parse native int!".to_string()))
+        }
+    } else if name.starts_with('\'') {
+        // Quoted! true or false?
+        match &name as &str {
+            "'null" => Ok(Value::Void),
+            "'true" => Ok(Value::Bool(true)),
+            "'false" => Ok(Value::Bool(false)),
+            _ => Err(Error::NotImplemented)
+        }
+    } else {
+        if let Some(value) = context.lookup_variable(name) {
+            Ok(value)
+        } else if let Some(value) = env.global_context.lookup_variable(name) {
+            Ok(value)
+        } else {
+            Err(Error::Undefined(format!("No such variable found in context: {}", name)))
+        }
+    }
+}
+
+pub fn lookup_function<'a> (name: &str, env: &Environment)-> Result<CallableType<'a>, Error> {
```

Comment:
From this code, it looks like the user will not be able to change the runtime behavior of a reserved function by creating a function with the same name.  However, I didn't see any code that prevents the user from doing so?

Reply:
Right -- reserved functions are used first, before a user defined function would be. However, we would want to generate a runtime error at the point of the `(define)` (and also, when the static analyzer is implemented, a static error). I started an issue #914.